### PR TITLE
[rfr] EmbargoTerminationApproval

### PR DESCRIPTION
**Pull request by samchrisinger**
:revolving_hearts: 
https://openscience.atlassian.net/browse/OSF-5641
:bouquet: :confetti_ball: :bouquet: 

# Purpose

We want a way for users to be able to end embargoed registrations public early. Borrowing from the patterns for Registrations, Embargoes, Retractions, etc, we will notify admins on the registration tree and request that they either approve or disapprove this action. If an admin fails to respond within 48hrs, they have granted implicit approval. If a user is the only project administrator, then the Embargo will be lifted immediately. 

# Changes

- Sanction and its subclasses are moved to their own module (website.project.sanctions)
- add new Sanction subclass EmbargoTerminationApproval
- add NodeLog.EMBARGO_TERMINATED/EMBARGO_TERMINATION_APPROVED/ EMBARGO_TERMINATION_CANCELLED and supporting templates (TODO)
- update nodesPrivacy.js/mako to support making Embargoes public
- move NodePrivacy instantiation into NodeControl, and pass onPrivacyChanged handler into NodePrivacy
- add tests for Node sanction states (is_pending_registration, is_embargoed, etc)

# Side Effects

The logic for restricting what can be made public/private is fairly scattered, so I wouldn't be surprised if some places cause issues (see: https://github.com/samchrisinger/osf.io/commit/0d1821dae05a24c2f5e1c2ea869ca1f837b0d2e8#diff-7a6d21a50624153d15e63a5f643401c7). Hopefully testing will resolve most or all of these.

Additionally this adds a new auto-approve script (like Embargoes and ReigstrationApprovals), that will needed to be added to celery-beat (?).

# Testing Notes

### Creating an Embargoed registration locally:
1. Create a project or navigate to an existing project
2. Create a registration of that project
  1. go to the 'Registrations' tab
  2. click 'New registration', pick 'Open Ended ...'
  3. Press 'Preview for submission'
  4. Press 'Register' and choose 'Enter registration into Embargo'; choose a valid end date
  5. Watch the log output of your `celery_worker`, it should dump an email that reads like:
```
[2016-04-11 13:47:24,410: DEBUG/Worker-4] Sending email...
[2016-04-11 13:47:24,411: DEBUG/Worker-4] To: sam@cos.io
From: openscienceframework-noreply@osf.io
Subject: Registration pending for one of your projects.
Message: Hello sam,

You initiated an embargoed registration of Another. The proposed registration can be viewed here: http://localhost:5000/x5agz/.

If approved, a registration will be created for the project and it will remain private until it is withdrawn, manually
made public, or the embargo end date has passed on 2016-04-30.

To approve this embargo, click the following link: http://localhost:5000/project/x5agz/?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhY3Rpb24iOiJhcHByb3ZlX2VtYmFyZ28iLCJzYW5jdGlvbl9pZCI6IjU3MGJlMzJiOWY4YjFmOGI1MTI0OWY3ZSIsInVzZXJfaWQiOiJ6cDduZCJ9.dXTcMHVXOgMrAkUQA3qrNZA9WlFyPkTzvBxSweOswew.

To cancel this embargo, click the following link: http://localhost:5000/project/2mh9z/?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhY3Rpb24iOiJyZWplY3RfZW1iYXJnbyIsInNhbmN0aW9uX2lkIjoiNTcwYmUzMmI5ZjhiMWY4YjUxMjQ5ZjdlIiwidXNlcl9pZCI6InpwN25kIn0.NouZZMjIiB5VocjzI3j287UejBTmyfY9QFQ-M0H_3bU.

Note: Clicking the disapproval link will immediately cancel the pending embargo and the
registration will remain in draft state. If you neither approve nor disapprove the embargo
within 48 hours from midnight tonight (EDT) the registration will remain
private and enter into an embargoed state.

Sincerely yours,

The OSF Robots
```

Copy the approve link, and paste it into your browser window

To do the same for a project with gt one admin, just copy/paste the approval link for each admin.

# TODOS
- [x] email templates
- [x] log templates
- [x] UI when embargo termination request is pending
- [x] tests and testing
- [x] EmbargoTerminationApproval auto-approve scripts
- [ ] celery-beat entry

# Screenshots

screen shot 2016-03-30 at 11 15 43 am
screen shot 2016-03-30 at 11 15 50 am
screen shot 2016-03-30 at 11 16 07 am
screen shot 2016-03-30 at 11 16 14 am

screen shot 2016-03-30 at 10 58 26 am

### CODE REVIEW

File: website/templates/log_templates.mako [hunk 1]
```diff
@@ -41,6 +41,27 @@ initiated an embargoed registration of
...
+
+<script type="text/html" id="embargo_terminated">
+ended the embargo for
+<a class="log-node-title-link overflow" data-bind="text: nodeTitle, attr: {href: nodeUrl}"></a>
+</script>
+
+<script type="text/html" id="embargo_terminated_no_user">
+Embargo for
+<a class="log-node-title-link overflow" data-bind="text: nodeTitle, attr: {href: nodeUrl}"></a>
+ended early.
```
Comment (samchrisinger) [inline website/templates/log_templates.mako]: - early

File: website/templates/log_templates.mako [hunk 2]
```diff
@@ -41,6 +41,27 @@ initiated an embargoed registration of
 <!-- /ko -->
 </script>
 
+<script type="text/html" id="embargo_termination_approved">
```
Comment (samchrisinger) [inline website/templates/log_templates.mako]: get rid of top three logs

File: website/templates/log_templates.mako [hunk 3]
```diff
@@ -41,6 +41,27 @@ initiated an embargoed registration of
...
+</script>
+
+<script type="text/html" id="embargo_terminated">
+ended the embargo for
+<a class="log-node-title-link overflow" data-bind="text: nodeTitle, attr: {href: nodeUrl}"></a>
+</script>
+
+<script type="text/html" id="embargo_terminated_no_user">
+Embargo for
+<a class="log-node-title-link overflow" data-bind="text: nodeTitle, attr: {href: nodeUrl}"></a>
```
Comment (samchrisinger) [inline website/templates/log_templates.mako]: should be registration URL

**Comment by whit537** [comment]
##### Punchlist—for @samchrisinger 

 - [ ] make Travis happy

**Comment by whit537** [comment]
##### Investigations—for @whit537

 - [ ] read code
 - [ ] run code

**Comment by samchrisinger** [comment]
@whit537 waiting on https://github.com/CenterForOpenScience/osf.io/pull/5371 to fix Travis woes

File: tests/test_models.py [hunk 4]
```diff
@@ -2601,12 +2600,12 @@ def test_remove_contributor(self):
 
     def test_manage_contributors_cannot_remove_last_admin_contributor(self):
         user2 = UserFactory()
-        self.project.add_contributor(contributor=user2, permissions=['read', 'write'], auth=self.auth)
+        self.project.add_contributor(contributor=user2, permissions=[READ, WRITE], auth=self.auth)
```
Comment (whit537) [inline tests/test_models.py]: :+1: +1

File: website/templates/emails/pending_embargo_termination_non_admin.txt.mako [hunk 5]
```diff
@@ -0,0 +1,9 @@
+Hello ${user.fullname},
+
+We just wanted to let you know that ${initiated_by} has initiated a request to end the embargo for a registration of ${project_name}. That registration can be viewed here: ${registration_link}${context.get('custom_message', '')}.
+
+If approved, the embargo will be ened and the registration and all of its components will be make public immediately.
```
Comment (whit537) [inline website/templates/emails/pending_embargo_termination_non_admin.txt.mako]: typo: `ened`

**Comment by whit537** [comment]
Having some trouble standing this up. Starting from a fresh database, checkout, and assets build: I login and I get a blank `/` and a blank `/myprojects/`. I see a 404 in the console that might be a clue ...

screen shot 2016-04-05 at 9 10 25 am

screen shot 2016-04-05 at 9 10 29 am

screen shot 2016-04-05 at 9 10 38 am

screen shot 2016-04-05 at 9 11 15 am

**Comment by whit537** [comment]
My expectation was that I'd be able to add projects from `/myprojects/`. Hmm ...

**Comment by chrisseto** [comment]
@whit537 Are you running webpack locally? It looks like you might be loading an older version of the dashboard. The `/v1/dashboard/*`  route(s) no longer exist.

**Comment by whit537** [comment]
> Are you running webpack locally?

I *thought* I was ...

> It looks like you might be loading an older version of the dashboard. The `/v1/dashboard/*` route(s) no longer exist.

Hmm, okay, let's see here ...

**Comment by samchrisinger** [comment]
+1 to chrisseto's comment. Perhaps run

`inv clean`
and make sure 
`inv assets -dw`
is running

**Comment by whit537** [comment]
> `inv clean`

Hmm ... that just removes `*.pyc` files (which I suppress globally with `PYTHONDONTWRITEBYTECODE`). `inv clean_assets` on the other hand ...

**Comment by whit537** [comment]
Yessssssssss!

Okay! Thanks for the tip. :-) Proceeding with review ...

screen shot 2016-04-05 at 12 08 12 pm

**Comment by whit537** [comment]
Okay, now with 6188fda I get a `SyntaxError` inside Mithril. Hmm ...

screen shot 2016-04-07 at 2 22 47 pm

screen shot 2016-04-07 at 2 24 36 pm

**Comment by chrisseto** [comment]
@whit537 Its probably an exception from the API as HTML. Mithril is trying to parse it as JSON.

**Comment by whit537** [comment]
Thanks.

**Comment by whit537** [comment]
Alright, I threw a breakpoint (using the Chrome dev tools) [in mithril](https://github.com/lhorie/mithril.js/blob/v0.2.0/mithril.js#L1127). I hit it three times when loading `/myprojects/` ... LOL

Okay, this is getting ridiculous. :-)

One one run through the breakpoints, `/myprojects/` loaded (more or less?) as expected. And then on the next ...

screen shot 2016-04-07 at 3 02 24 pm

**Comment by samchrisinger** [comment]
@whit537 seems like your HD may be close to full.

There are no changes introduced into the dashboard/myprojects code in the PR.

**Comment by whit537** [comment]
> seems like your HD may be close to full.

Right. I guess 5.2 Gi isn't enough. :(

```
$ df -h
Filesystem      Size   Used  Avail Capacity  iused   ifree %iused  Mounted on
/dev/disk1     112Gi  106Gi  5.2Gi    96% 27949960 1371766   95%   /
devfs          191Ki  191Ki    0Bi   100%      660       0  100%   /dev
map -hosts       0Bi    0Bi    0Bi   100%        0       0  100%   /net
map auto_home    0Bi    0Bi    0Bi   100%        0       0  100%   /home
$
```

> There are no changes introduced into the dashboard/myprojects code in the PR.

I know, that's why this is so weird. I don't believe I've touched the OSF since the last time I was here on this PR.

I'll stop posting here to keep the noise down ...

**Comment by whit537** [comment]
I am unfrobulated (#5419). Sorry for the distraction, @samchrisinger @chrisseto. I guess the lesson is "check your local.py". Okay! Back on track here ...

**Comment by whit537** [comment]
I can create am embargoed registration. This PR is about lifting embargoes early, I guess ...

screen shot 2016-04-07 at 5 22 34 pm
screen shot 2016-04-07 at 5 35 17 pm
screen shot 2016-04-07 at 5 35 40 pm

**Comment by whit537** [comment]
Made the project public, not seeing the embargo notice ... because I'm still on aa1042c. ;-)

screen shot 2016-04-07 at 5 38 32 pm

**Comment by whit537** [comment]
Alright, I'm on 6188fda and still not seeing the "End Embargo Early" notice, gonna have to try harder ... maybe it's a clue that the registrations are still in draft even though I thought I registered them?

Oh well, more tomorrow. :-)

screen shot 2016-04-07 at 5 52 13 pm

**Comment by whit537** [comment]
I'm trying to get a registration out of "Draft Registrations" and into "Registrations."

screen shot 2016-04-08 at 4 48 01 pm

**Comment by whit537** [comment]
I'm now trying with an open-ended registration and no embargo, just to get *something* working. When I "Continue" on the registration choice dialog I see my new registration listed under "Registrations," but then when I refresh it is gone. Maybe something with file copying failing?

screen shot 2016-04-08 at 4 55 54 pm

**Comment by whit537** [comment]
Hmmm ... I left it a long time, and then refreshed. Now it's still there, but I can't find anything to do with it. Nothing is clickable.

screen shot 2016-04-08 at 5 43 52 pm

**Comment by whit537** [comment]
Eep! Now it's gone again. :disappointed:

**Comment by whit537** [comment]
All these drafts have a disabled "Register" button.

screen shot 2016-04-08 at 5 45 32 pm

**Comment by whit537** [comment]
I'm sorry to be so dense, @samchrisinger. Can you give me the steps to create an approved, embargoed registration? It looks from [the Jira description](https://openscience.atlassian.net/browse/OSF-5641) that that's the starting point for the functionality on this PR, but I don't know how to get there. :disappointed:

**Comment by whit537** [comment]
Digging into this some more ...

**Comment by samchrisinger** [comment]
@whit537 sorry, haven't had a ton of time to respond to this. 

Are you running rabbitmq and celery locally?

**Comment by whit537** [comment]
> Are you running rabbitmq and celery locally?

Yes, I am.

**Comment by whit537** [comment]
I find some "registration" collections in Mongo ...

screen shot 2016-04-11 at 11 04 29 am

**Comment by whit537** [comment]
> sorry, haven't had a ton of time to respond to this.

Maybe I will reach out to @csoderberg, she gave me my first product orientation to OSF. Perhaps she can help me understand the registration and embargo system.

@csoderberg Up for a Skype sometime today or tomorrow? :-)

**Comment by whit537** [comment]
I've also reached out [on Jira](https://openscience.atlassian.net/browse/OSF-5641).

File: api/base/exceptions.py [hunk 6]
```diff
@@ -216,7 +216,7 @@ class TwoFactorRequiredError(AuthenticationFailed):
 
 class InvalidModelValueError(JSONAPIException):
     status_code = 400
-    default_detail = _('Invalid value in POST/PUT/PATCH request.')
+    default_detail = ('Invalid value in POST/PUT/PATCH request.')
```
Comment (samchrisinger) [inline api/base/exceptions.py]: Undo this

**Comment by csoderberg** [comment]
@whit537 responded on flowdock with options

**Comment by samchrisinger** [comment]
@whit537 see updated PR description. No problem creating approved embargoed registrations on my end.

**Comment by whit537** [comment]
Okay, so Step 2.v is where I'm hung up, and the message I get in the UI suggests that a malfunctioning waterbutler may be in teh way of me and my email. Reopening https://github.com/CenterForOpenScience/osf.io/issues/5419 to investigate ...

screen shot 2016-04-11 at 3 54 10 pm

> Files are being copied to the newly created registration, and you will receive an email notification when the copying is finished.

**Comment by whit537** [comment]
Okay, reunfrobulated! I have an approved embargoed registration!

screen shot 2016-04-11 at 4 15 42 pm

**Comment by whit537** [comment]
And making the registration public does give me the "End Embargo Early" dialog, which does appear to work.

screen shot 2016-04-11 at 4 20 06 pm
screen shot 2016-04-11 at 4 21 25 pm

**Comment by whit537** [comment]
Running through with multiple admins ...

**Comment by whit537** [comment]
Sweet mercy. I'm really sloggin' through the mud here, aren't I? 😳  😩 

I signed up a second contributor in a different browser session, and when I try to add the second contributor from the first session I get a 500 from the search endpoint.

```
[werkzeug]  INFO: 127.0.0.1 - - [11/Apr/2016 15:39:07] "GET /api/v1/user/search/?query=bar&page=0&_=1460407105699 HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/site-packages/flask/app.py", line 1836, in __call__
    return self.wsgi_app(environ, start_response)
  File "/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/site-packages/flask/app.py", line 1820, in wsgi_app
    response = self.make_response(self.handle_exception(e))
  File "/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/site-packages/flask/app.py", line 1403, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/site-packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/site-packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/site-packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/whit537/workbench/cos/osf.io/env/lib/python2.7/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/whit537/workbench/cos/osf.io/framework/routing/__init__.py", line 118, in wrapped
    data = fn(*args, **kwargs)
  File "/Users/whit537/workbench/cos/osf.io/framework/auth/decorators.py", line 22, in wrapped
    return func(*args, **kwargs)
  File "/Users/whit537/workbench/cos/osf.io/website/search/views.py", line 208, in search_contributor
    exclude=exclude, current_user=user)
  File "/Users/whit537/workbench/cos/osf.io/website/search/search.py", line 19, in wrapped
    return func(*args, **kwargs)
  File "/Users/whit537/workbench/cos/osf.io/website/search/search.py", line 89, in search_contributor
    exclude=exclude, current_user=current_user)
  File "/Users/whit537/workbench/cos/osf.io/website/search/elastic_search.py", line 79, in wrapped
    return func(*args, **kwargs)
  File "/Users/whit537/workbench/cos/osf.io/website/search/elastic_search.py", line 593, in search_contributor
    if current_user and current_user._id == user._id:
AttributeError: 'NoneType' object has no attribute '_id'
```

**Comment by samchrisinger** [comment]
@whit537 maybe try running `inv rebuild_search`

**Comment by whit537** [comment]
Still not out of the woods on emails getting spit out properly. This system is feeling pretty brittle to me right about now. :-(

Done for the day, more later ...

screen shot 2016-04-11 at 5 01 21 pm

**Comment by whit537** [comment]
They `pyrax` import in that script is [a year old](https://github.com/CenterForOpenScience/osf.io/blame/e731c4c1598f2118195528f8b818bd99dfc7dcad/scripts/osfstorage/files_audit.py#L20).

**Comment by whit537** [comment]
(Sorry, wrong thread ...)

File: website/static/css/koHelpers.css [hunk 7]
```diff
@@ -0,0 +1,3 @@
+.ensure-bs-tooltips {
+    pointer-events: auto !important;
+}
```
Comment (mfraezz) [inline website/static/css/koHelpers.css]: screen shot 2016-04-20 at 16 37 52
ლ(ಠ益ಠლ)

**Comment by mfraezz** [comment]
This is going into `release/0.68.0`, but it has a (relatively small) merge conflict:
`CONFLICT (content): Merge conflict in website/project/views/node.py`
```python
<<<<<<< HEAD
        'archiving': node.archiving or getattr(node.root, 'archiving', False),
=======
        'is_embargoed': node.is_embargoed,
        'archiving': node.archiving,
>>>>>>> ee682f9bb28ffde4d11fc8d6677e07cc23aa0476
```

**Comment by mfraezz** [comment]
Behavior for single-admin embargoes looks right, but multi-admin behavior is broken -- emails are never sent out, API throws a `500` when building emails:

```
Traceback (most recent call last):
  [DRF trace omitted]
  File "/Users/matt/Devel/osf.io/api/base/serializers.py", line 871, in update
    ret['data'].append(self.child.update(resource, data))
  File "/Users/matt/Devel/osf.io/api/nodes/serializers.py", line 258, in update
    node.update(validated_data, auth=auth)
  File "/Users/matt/Devel/osf.io/website/project/model.py", line 1429, in update
    save=False
  File "/Users/matt/Devel/osf.io/website/project/model.py", line 3084, in set_privacy
    self.request_embargo_termination(auth=auth)
  File "/Users/matt/Devel/osf.io/website/project/model.py", line 3447, in request_embargo_termination
    approval.ask(admins)
  File "/Users/matt/Devel/osf.io/website/project/sanctions.py", line 249, in ask
    self._notify_authorizer(contrib, node)
  File "/Users/matt/Devel/osf.io/website/project/sanctions.py", line 317, in _notify_authorizer
    context = self._email_template_context(authorizer, node, is_authorizer=True)
  File "/Users/matt/Devel/osf.io/website/project/sanctions.py", line 953, in _email_template_context
    context = super(EmbargoTerminationApproval, self)._email_template_context(user, node, is_authorizer, urls)
TypeError: _email_template_context() takes at most 4 arguments (5 given)
[20/Apr/2016 16:24:17] "PATCH /v2/nodes/ HTTP/1.1" 500 59
```

File: website/static/js/nodesPrivacy.js [hunk 8]
```diff
@@ -165,114 +182,185 @@ var NodesPrivacyViewModel = function(parentIsPublic) {
...
+        if (res.data[0].attributes.public) {
+            // TODO move all logic to onSetPrivacy handler and don't reload the page.
+            // Requires the components list to be rendered client-side.
+            window.location.reload();
+        }
+        else {
+            self.onSetPrivacy(nodesChanged, true);
+            $osf.growl(
+                'Request Initiated',
+                'You have initiated a request to end this registration\'s embargo early, and to make it and all of its components public immediately. All adminstrators on this registration have 48 hours to approve or disapprove of this acrion.',
```
Comment (mfraezz) [inline website/static/js/nodesPrivacy.js]: >48 hours to approve or disapprove of this **acrion**.

**Comment by mfraezz** [comment]
Following either link in the multi-admin case also doesn't work:

screen shot 2016-04-20 at 17 58 53
screen shot 2016-04-20 at 17 59 16

**Comment by mfraezz** [comment]
Pass complete. Aside from the above bugs, looks good. Please address those issues and add a few tests for a multi-admin embargo cancellation case. Should be RTM soon :fire:

**Comment by mfraezz** [comment]
Emails are now working, but when I follow the approval link as the second admin, I get... mixed signals:

screen shot 2016-04-22 at 10 35 00

**Comment by mfraezz** [comment]
Why do two emails go out when there are two admins, but zero are sent when there's one admin?

File: tests/test_registrations/test_archiver.py [hunk 9]
```diff
@@ -1130,7 +1130,7 @@ def test_archiving_nodes_not_added_to_search_on_archive_failure(self, mock_send,
                 mock.patch('website.archiver.model.ArchiveJob.success', mock.PropertyMock(return_value=False))
         ) as (mock_finished, mock_sent, mock_success):
             listeners.archive_callback(reg)
-        mock_update_search.assert_not_called()
+        mock_delete_index_node.assert_called()
```
Comment (mfraezz) [inline tests/test_registrations/test_archiver.py]: I'm not certain that this test is testing anything. `.assert_called` isn't in the [Mock docs](https://docs.python.org/3/library/unittest.mock.html#the-mock-class), and there's a [great blog post](http://engineeringblog.yelp.com/2015/02/assert_called_once-threat-or-menace.html) about why you should be very careful when calling methods on `MagicMocks` that look like they might be asserting things. Excerpt: 

>Here’s a clue: the test will happily continue to pass if you change the assertion to:
>
>`mock_restart.assert_two_wrongs_dont_make_a_right()`
>
>Or:
>
>`mock_restart.assert_two_plus_two_equals(5)`
>
>Or even:
>
>`mock_restart.ASDFASDFASDFASDFASDFASDFASDF()`
>
>What’s going on here?
>
>Remember that a mock’s job is to say, “You got it, boss” whenever anyone calls it.

Comment (mfraezz) [inline tests/test_registrations/test_archiver.py]: Output from the `MockChecker` at the bottom of the blog post:

```
~/D/osf.io ❯ python -m scripts.mock_sniffer tests/*.py tests/*/*.py                                                                                  ⏎ remotes/origin/pr/5294 ✭ ◼
tests/test_addons_oauth.py:131: you may have called a nonexistent method on mock
tests/test_cas_authentication.py:170: you may have called a nonexistent method on mock
tests/test_models.py:3112: you may have called a nonexistent method on mock
tests/test_models.py:3118: you may have called a nonexistent method on mock
tests/test_models.py:3149: you may have called a nonexistent method on mock
tests/test_models.py:3165: you may have called a nonexistent method on mock
tests/test_utils.py:409: you may have called a nonexistent method on mock
tests/test_utils.py:415: you may have called a nonexistent method on mock
tests/test_views.py:2323: you may have called a nonexistent method on mock
tests/test_views.py:2377: you may have called a nonexistent method on mock
tests/test_views.py:2399: you may have called a nonexistent method on mock
tests/test_views.py:3305: you may have called a nonexistent method on mock
tests/test_views.py:3321: you may have called a nonexistent method on mock
tests/webtest_tests.py:279: you may have called a nonexistent method on mock
tests/webtest_tests.py:298: you may have called a nonexistent method on mock
tests/framework_tests/test_email.py:50: you may have called a nonexistent method on mock
tests/framework_tests/test_sentry.py:74: you may have called a nonexistent method on mock
tests/models/test_user.py:496: you may have called a nonexistent method on mock
tests/test_registrations/test_archiver.py:336: you may have called a nonexistent method on mock
tests/test_registrations/test_archiver.py:848: you may have called a nonexistent method on mock
tests/test_registrations/test_archiver.py:856: you may have called a nonexistent method on mock
tests/test_registrations/test_archiver.py:880: you may have called a nonexistent method on mock
tests/test_registrations/test_archiver.py:882: you may have called a nonexistent method on mock
tests/test_registrations/test_archiver.py:925: you may have called a nonexistent method on mock
tests/test_registrations/test_archiver.py:942: you may have called a nonexistent method on mock
tests/test_registrations/test_archiver.py:1023: you may have called a nonexistent method on mock
tests/test_registrations/test_archiver.py:1028: you may have called a nonexistent method on mock
tests/test_registrations/test_archiver.py:1033: you may have called a nonexistent method on mock
tests/test_registrations/test_archiver.py:1103: you may have called a nonexistent method on mock
tests/test_registrations/test_archiver.py:1119: you may have called a nonexistent method on mock
tests/test_registrations/test_archiver.py:1133: you may have called a nonexistent method on mock
tests/test_registrations/test_archiver.py:1143: you may have called a nonexistent method on mock
tests/test_registrations/test_embargoes.py:405: you may have called a nonexistent method on mock
tests/test_registrations/test_models.py:192: you may have called a nonexistent method on mock
tests/test_registrations/test_registration_approvals.py:259: you may have called a nonexistent method on mock
tests/test_registrations/test_sanctions.py:153: you may have called a nonexistent method on mock
tests/test_registrations/test_sanctions.py:154: you may have called a nonexistent method on mock
tests/test_registrations/test_sanctions.py:185: you may have called a nonexistent method on mock
tests/test_registrations/test_sanctions.py:186: you may have called a nonexistent method on mock
tests/test_registrations/test_sanctions.py:194: you may have called a nonexistent method on mock
tests/test_registrations/test_sanctions.py:195: you may have called a nonexistent method on mock
tests/test_registrations/test_sanctions.py:246: you may have called a nonexistent method on mock
tests/test_registrations/test_sanctions.py:258: you may have called a nonexistent method on mock
```

Comment (mfraezz) [inline tests/test_registrations/test_archiver.py]: You obviously don't need to fix all of these, but a few are sneaking in in this PR. Please don't let them.

Comment (samchrisinger) [inline tests/test_registrations/test_archiver.py]: @mfraezz we should confer with @sloria and discuss adding this as a hard check in our Travis builds.

**Comment by samchrisinger** [comment]
Closing, will reopen against 0.68

File: website/static/js/nodesPrivacy.js [hunk 10]
```diff
@@ -18,7 +18,8 @@ var MESSAGES = {
     'Please review your projects, components, and add-ons for sensitive or restricted information before making them public.' +
     '<br><br>Once they are made public, you should assume they will always be public. You can ' +
     'return them to private later, but search engines (including Google’s cache) or others may access files before you do.',
-
+    makeEmbargoPublicWarning: 'By making this registration public, you will end the embargo period and the registration will be open to the public. Making this registration public will automatically make all of its components public as well. This action is irreversible.',
+    makeEmbargoPublicTitle: 'End Embargo Early',
```
Comment (sloria) [inline website/static/js/nodesPrivacy.js]: As per osf-style, modal titles should be sentence-case.

File: website/static/js/nodeControl.js [hunk 11]
```diff
@@ -265,6 +274,21 @@ function NodeControl (selector, data, options) {
...
+        new NodesPrivacy('#nodesPrivacy', self.data.node, function(nodesChanged, requestedEmbargoTermination) {
+            // TODO: The goal here is to update the UI of the project dashboard to
+            // reflect the new privacy state(s). Unfortunately, since the components
+            // view is rendered server-side we have a relatively limited capacity to
+            // update the page. Once the components list is componentized we can
+            // rerender just that section of the DOM (as well as updating the
+            // project-specific UI).
+            // For now, this method only needs to handle requests for making embargoed
+            // nodes public.
+            if (requestedEmbargoTermination) {
```
Comment (sloria) [inline website/static/js/nodeControl.js]: I'm a bit confused why this needs to happen here, and not within the `NodesPrivacyViewModel`. Can you clarify?

Comment (sloria) [inline website/static/js/nodeControl.js]: Oohh, disregard that. I now realize that `self.viewModel` is `ProjectViewModel`, not `NodesPrivacyViewModel`.

Comment (samchrisinger) [inline website/static/js/nodeControl.js]: The KO view-model scope disallows programmatically changing the state of the Private/Public button. Rather than using $ or equivalent to do this, I chose to pass a callback into NodePrivacy.

From an architecture standpoint, it seems to me that since NodesPrivacy is a subcomponent of NodeControl the NodeControl should handle modifications to the Node and update the state of the child view model as appropriate.

File: website/static/js/nodesPrivacy.js [hunk 12]
```diff
@@ -165,114 +182,185 @@ var NodesPrivacyViewModel = function(parentIsPublic) {
...
+            nodesState[node].public = false;
+            nodesState[node].changed = nodesState[node].public !== this.nodesOriginal[node].public;
+
+        }
+    }
+    this.nodesState(nodesState);
+    m.redraw(true);
+};
+
+NodesPrivacyViewModel.prototype.back = function() {
```
Comment (sloria) [inline website/static/js/nodesPrivacy.js]: +0 on moving all methods to the prototype. The memory savings we get from assigning to `prototype` is close to nil, since we almost always instantiate a ViewModel only once. Probably not even worth the extra typing (though I suppose we could use `oop.defclass` to reduce boilerplate), but this is fine as is.

Comment (samchrisinger) [inline website/static/js/nodesPrivacy.js]: Mostly just a sanity check for me as I prepared to do refactor in the module. Agreed memory savings are not worth it, but for readability I vastly prefer prototype methods defined outside of the constructor. :shrug:

File: website/templates/emails/pending_embargo_termination_admin.txt.mako [hunk 13]
```diff
@@ -0,0 +1,14 @@
+Hello ${user.fullname},
+
+% if is_initiator:
+You initiated a request to end the embargo for a registration of ${project_name}${context.get('custom_message', '')}. The embargoed registration can be viewed here: ${registration_link}.
```
Comment (sloria) [inline website/templates/emails/pending_embargo_termination_admin.txt.mako]: Can be more concise: "You requested to end..."

File: website/templates/emails/pending_embargo_termination_non_admin.txt.mako [hunk 14]
```diff
@@ -0,0 +1,9 @@
+Hello ${user.fullname},
+
+We just wanted to let you know that ${initiated_by} has initiated a request to end the embargo for a registration of ${project_name}. That registration can be viewed here: ${registration_link}${context.get('custom_message', '')}.
```
Comment (sloria) [inline website/templates/emails/pending_embargo_termination_non_admin.txt.mako]: Same here: "We just wanted to let you know that ... has requested to end..."

File: website/templates/emails/pending_embargo_termination_admin.txt.mako (hunk 13 above)
Comment (samchrisinger) [inline website/templates/emails/pending_embargo_termination_admin.txt.mako]: Hate the player not the game: https://openscience.atlassian.net/browse/OSF-5641

File: website/util/time.py [hunk 15]
```diff
@@ -0,0 +1,16 @@
+# -*- coding: utf-8 -*-
```
Comment (sloria) [inline website/util/time.py]: This is unrelated code; possibly the result of the cherry-pick? Can these commits be easily removed?

**Comment by sloria** [comment]
Ah, gee, I just realized I've been commenting on the wrong PR. Disregard the comment re: the unrelated code.

File: website/templates/emails/pending_embargo_termination_admin.txt.mako (hunk 13 above)
Comment (sloria) [inline website/templates/emails/pending_embargo_termination_admin.txt.mako]: Ha, OK, this is NBD; this is fine as is.

File: website/static/js/nodeControl.js (hunk 11 above)
Comment (sloria) [inline website/static/js/nodeControl.js]: Yep, you're right; I realized my misunderstanding of this right after I posted my comment.
